Source/Core/Duality/Utility/Coroutines/Coroutine.cs

```diff
@@ -0,0 +1,97 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using Duality.Resources;
+
+namespace Duality.Utility.Coroutines
+{
+	/// <summary>
+	/// A Coroutine's management class
+	/// </summary>
+	public sealed class Coroutine
+	{
+		private IEnumerator<WaitUntil> enumerator = null;
+		private WaitUntil currentCondition;
+
+		/// <summary>
+		/// The Coroutine's current <see cref="CoroutineStatus">execution status</see>
+		/// </summary>
+		public CoroutineStatus Status { get; private set; }
+		/// <summary>
+		/// The Exception encountered during the last frame
+		/// </summary>
+		public Exception LastException { get; private set; }
+		/// <summary>
+		/// The Coroutine's Name
+		/// </summary>
+		public string Name { get; private set; }
+		/// <summary>
+		/// True if this Coroutine's Status is <see cref="CoroutineStatus.Paused"/> or <see cref="CoroutineStatus.Running"/>
+		/// </summary>
+		public bool IsAlive
+		{
+			get { return this.Status == CoroutineStatus.Paused || this.Status == CoroutineStatus.Running; }
+		}
+
+		internal void Setup(IEnumerable<WaitUntil> values, string name)
+		{
+			this.Status = CoroutineStatus.Running;
+
+			this.LastException = null;
+			this.Name = name;
```

Comment:
For the name of the coroutine (which is used only for logging I believe) I think we can simply use `enumerator.GetType().FullName`. The name will turn into something like `Duality.Tests.Utility.CoroutinesTest+<BasicRoutine>d__5`. Not only does it gets the name automatically now we can also postpone doing this till we actually need to log it so we can also remove the `Name` property as well making `Coroutine` a bit smaller/faster.

Reply:
but you'd lose the possibility to give it a meaningful name, i.e. you spawn a bunch of coroutines, with the same method, but that affect a different object that you provide as a parameter.

With a proper name, you could track down the issue to the exact instance even without the need to debugging.

I could do `this.Name = name ?? this.enumerator.GetType().FullName` and move it after the enumerator reassignment.

Reply:
If you go that far to give it such a detailed name wouldn't it be more practical to just add in logging to the coroutine itself? We are already logging the exception and that exception should carry this information if its important.

I think in most cases users just don't bother assigning a readable name and keeping them in sync.